tendermint/src/lite/verifier.rs

```diff
@@ -184,98 +124,120 @@ where
     SH: SignedHeader<Commit = C>,
     C: Commit,
     L: TrustThreshold,
-    S: Store<TrustedState = TS>,
-    R: Requester<SignedHeader = SH, ValidatorSet = C::ValidatorSet>,
 {
-    // can we trust the still untrusted header based on the given trusted state?
-    match check_support(
-        trusted_state,
+    // ensure the new height is higher
+
+    // validate the untrusted header against its commit, vals, and next_vals
+    let untrusted_header = untrusted_sh.header();
+    let untrusted_commit = untrusted_sh.commit();
+    validate_signed_header_and_vals(
         untrusted_header,
-        trust_threshold,
-        trusting_period,
-        now,
-    ) {
-        Ok(_) => {
-            let untrusted_next_vals =
-                req.validator_set(untrusted_header.header().height().increment())?;
-            let untrusted_state = TS::new(untrusted_header, &untrusted_next_vals);
-            store.add(&untrusted_state)?;
-            return Ok(());
-        }
-        Err(e) => {
-            if e != Error::InsufficientVotingPower {
-                return Err(e);
+        untrusted_commit,
+        untrusted_vals,
+        untrusted_next_vals,
+    )?;
+
+    // ensure the new height is higher.
+    // if its +1, ensure the vals are correct.
+    // if its >+1, ensure we can skip to it
+    let trusted_header = trusted_state.last_header().header();
+    let trusted_height = trusted_header.height();
+    let untrusted_height = untrusted_sh.header().height();
+    match untrusted_height.cmp(&trusted_height.increment()) {
+        Ordering::Less => return Err(Error::NonIncreasingHeight),
+        Ordering::Equal => {
+            let trusted_vals_hash = trusted_header.next_validators_hash();
+            let untrusted_vals_hash = untrusted_header.validators_hash();
+            if trusted_vals_hash != untrusted_vals_hash {
+                // TODO: more specific error
+                // ie. differentiate from when next_vals.hash() doesnt
+                // match the header hash ...
+                return Err(Error::InvalidNextValidatorSet);
             }
         }
+        Ordering::Greater => {
+            let trusted_vals = trusted_state.validators();
+            verify_commit_trusting(trusted_vals, untrusted_commit, trust_threshold)?;
+        }
     }
 
-    // Here we can't trust the passed in untrusted header based on the known trusted state.
-    // Run bisection: try again with a pivot header whose height is in the
-    // middle of the trusted height and the desired height (h2.height).
-
-    let trusted_header = trusted_state.last_header();
-    let trusted_height: u64 = trusted_header.header().height().value();
-    let untrusted_height: u64 = untrusted_header.header().height().value();
-    let pivot: u64 = (trusted_height + untrusted_height) / 2;
-    let pivot_header = req.signed_header(pivot)?;
-    let pivot_vals = req.validator_set(pivot)?;
+    // verify the untrusted commit
```

Comment:
```suggestion
```
Comment doesn't add more context.

Reply:
What about:

```suggestion
    // All validation passed successfully. Verify the validators correctly committed the block.
```

? Or you think not needed at all?

Reply:
Your suggestion does indeed give more context and should be kept.